com.microsoft.java.debug.core/src/main/java/com/microsoft/java/debug/core/Watchpoint.java

```diff
@@ -0,0 +1,247 @@
+        });
+        requests.clear();
+        subscriptions.clear();
+    }
+
+    @Override
+    public String className() {
+        return className;
+    }
+
+    @Override
+    public String fieldName() {
+        return fieldName;
+    }
+
+    @Override
+    public String accessType() {
+        return accessType;
+    }
+
+    @Override
+    public String getCondition() {
+        return condition;
+    }
+
+    @Override
+    public void setCondition(String condition) {
+        this.condition = condition;
+        setCompiledConditionalExpression(null);
+    }
+
+    @Override
+    public int getHitCount() {
+        return hitCount;
+    }
+
+    @Override
+    public void setHitCount(int hitCount) {
+        this.hitCount = hitCount;
+
+        Observable.fromIterable(this.requests())
+            .filter(request -> request instanceof WatchpointRequest)
+            .subscribe(request -> {
+                request.addCountFilter(hitCount);
+                request.enable();
+            });
+    }
+
+    @Override
+    public void putProperty(Object key, Object value) {
+        propertyMap.put(key, value);
+    }
+
+    @Override
+    public Object getProperty(Object key) {
+        return propertyMap.get(key);
+    }
+
+    @Override
+    public CompletableFuture<IWatchpoint> install() {
+        // It's possible that different class loaders create new class with the same name.
+        // Here to listen to future class prepare events to handle such case.
+        ClassPrepareRequest classPrepareRequest = vm.eventRequestManager().createClassPrepareRequest();
+        classPrepareRequest.addClassFilter(className);
+        classPrepareRequest.enable();
+        requests.add(classPrepareRequest);
+
+        CompletableFuture<IWatchpoint> future = new CompletableFuture<>();
+        Disposable subscription = eventHub.events()
+            .filter(debugEvent -> debugEvent.event instanceof ClassPrepareEvent && (classPrepareRequest.equals(debugEvent.event.request())))
+            .subscribe(debugEvent -> {
+                ClassPrepareEvent event = (ClassPrepareEvent) debugEvent.event;
+                List<WatchpointRequest> watchpointRequests = createWatchpointRequests(event.referenceType());
+                requests.addAll(watchpointRequests);
+                if (!watchpointRequests.isEmpty() && !future.isDone()) {
+                    this.putProperty("verified", true);
+                    future.complete(this);
+                }
+            });
+        subscriptions.add(subscription);
+
+        List<EventRequest> watchpointRequests = new ArrayList<>();
+        List<ReferenceType> types = vm.classesByName(className);
+        for (ReferenceType type : types) {
+            watchpointRequests.addAll(createWatchpointRequests(type));
+        }
+
+        requests.addAll(watchpointRequests);
+        if (!watchpointRequests.isEmpty() && !future.isDone()) {
+            this.putProperty("verified", true);
```

Comment:
this.putProperty("verified", true); twice ? see line 159

Reply:
By design. It's guarded by the condition check `!future.isDone()`

Reply:
List<WatchpointRequest> watchpointRequests = createWatchpointRequests(event.referenceType());
                requests.addAll(watchpointRequests);
                if (!watchpointRequests.isEmpty() && !future.isDone()) {
                    this.putProperty("verified", true);
                    future.complete(this);
                }
duplicate